// @yamada-ui/utils/dist/calc.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/calc.ts
var calc_exports = {};
__export(calc_exports, {
  calc: () => calc
});
module.exports = __toCommonJS(calc_exports);
function toExpression(operator, ...args) {
  return args.join(` ${operator} `).replace(/calc/g, "");
}
function add(...args) {
  return `calc(${toExpression("+", ...args)})`;
}
function subtract(...args) {
  return `calc(${toExpression("-", ...args)})`;
}
function multiply(...args) {
  return `calc(${toExpression("*", ...args)})`;
}
function divide(...args) {
  return `calc(${toExpression("/", ...args)})`;
}
function negate(value) {
  if (!isNaN(parseFloat(value.toString())))
    return String(value).startsWith("-") ? String(value).slice(1) : `-${value}`;
  return multiply(value, -1);
}
var calc = Object.assign(
  (x) => ({
    add: (...args) => calc(add(x, ...args)),
    divide: (...args) => calc(divide(x, ...args)),
    multiply: (...args) => calc(multiply(x, ...args)),
    negate: () => calc(negate(x)),
    subtract: (...args) => calc(subtract(x, ...args)),
    toString: () => x.toString()
  }),
  {
    add,
    divide,
    multiply,
    negate,
    subtract
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  calc
});
