// @yamada-ui/utils/dist/index.mjs
import {
  interopDefault
} from "./chunk-APJY76CK.mjs";
import {
  clampNumber,
  countDecimal,
  percentToValue,
  roundNumberToStep,
  toNumber,
  toPrecision,
  valueToPercent
} from "./chunk-7XRVZGET.mjs";
import {
  assignRef,
  createContext,
  createId,
  cx,
  findChild,
  findChildren,
  getValidChildren,
  includesChildren,
  isRefObject,
  isSomeElement,
  isValidElement,
  mergeRefs,
  omitChildren,
  pickChildren,
  useAsync,
  useAsyncFunc,
  useAsyncRetry,
  useCallbackRef,
  useIsMounted,
  useMergeRefs,
  useSafeLayoutEffect,
  useUnmountEffect,
  useUpdateEffect
} from "./chunk-NVC6I45P.mjs";
import {
  antonym,
  escape,
  toCamelCase,
  toKebabCase,
  toPascalCase,
  toTitleCase,
  transformSize
} from "./chunk-VCO3C2KN.mjs";
import {
  filterEmpty
} from "./chunk-5EG6NSMV.mjs";
import {
  calc
} from "./chunk-5J34UVTP.mjs";
import {
  COLOR_SCHEMES,
  SEMANTIC_COLOR_SCHEMES,
  TONES,
  alphaToHex,
  calcFormat,
  convertColor,
  darkenColor,
  getAlpha,
  getColor,
  hslaTo,
  hsvTo,
  isAccessible,
  isDark,
  isGray,
  isLight,
  isTone,
  lightenColor,
  parseToHsla,
  parseToHsv,
  parseToRgba,
  randomColor,
  rgbaTo,
  sameColor,
  shadeColor,
  tintColor,
  transparentizeColor
} from "./chunk-YREUKSHW.mjs";
import {
  assignAfter,
  filterObject,
  filterUndefined,
  flattenObject,
  getMemoizedObject,
  getObject,
  keysFormObject,
  memoizeObject,
  merge,
  objectFromEntries,
  omitObject,
  pickObject,
  replaceObject,
  splitObject
} from "./chunk-EJPLJ6FG.mjs";
import {
  ariaAttr,
  createdDom,
  dataAttr,
  getActiveElement,
  getAllFocusable,
  getEventRelatedTarget,
  getOwnerDocument,
  getOwnerWindow,
  getPlatform,
  getPx,
  hasNegativeTabIndex,
  hasTabIndex,
  isActiveElement,
  isApple,
  isContains,
  isContentEditable,
  isDisabled,
  isElement,
  isFocusable,
  isHTMLElement,
  isHidden,
  isMac,
  isSafari,
  isTabbable,
  isTouchDevice,
  platform,
  vendor
} from "./chunk-47Z7BHBJ.mjs";
import {
  addDomEvent,
  addPointerEvent,
  getEventPoint,
  getEventWindow,
  isMouseEvent,
  isMultiTouchEvent,
  isTouchEvent,
  pointFromMouse,
  pointFromTouch
} from "./chunk-SAYAT2RG.mjs";
import {
  funcAll,
  handlerAll,
  noop,
  runIfFunc
} from "./chunk-23I7TQOH.mjs";
import {
  cast,
  is,
  isArray,
  isBoolean,
  isEmpty,
  isEmptyObject,
  isFunction,
  isNotNumber,
  isNull,
  isNumber,
  isNumeric,
  isObject,
  isString,
  isUndefined,
  isUnit
} from "./chunk-QWVYBC6R.mjs";
export {
  COLOR_SCHEMES,
  SEMANTIC_COLOR_SCHEMES,
  TONES,
  addDomEvent,
  addPointerEvent,
  alphaToHex,
  antonym,
  ariaAttr,
  assignAfter,
  assignRef,
  calc,
  calcFormat,
  cast,
  clampNumber,
  convertColor,
  countDecimal,
  createContext,
  createId,
  createdDom,
  cx,
  darkenColor,
  dataAttr,
  escape,
  filterEmpty,
  filterObject,
  filterUndefined,
  findChild,
  findChildren,
  flattenObject,
  funcAll,
  getActiveElement,
  getAllFocusable,
  getAlpha,
  getColor,
  getEventPoint,
  getEventRelatedTarget,
  getEventWindow,
  getMemoizedObject,
  getObject,
  getOwnerDocument,
  getOwnerWindow,
  getPlatform,
  getPx,
  getValidChildren,
  handlerAll,
  hasNegativeTabIndex,
  hasTabIndex,
  hslaTo,
  hsvTo,
  includesChildren,
  interopDefault,
  is,
  isAccessible,
  isActiveElement,
  isApple,
  isArray,
  isBoolean,
  isContains,
  isContentEditable,
  isDark,
  isDisabled,
  isElement,
  isEmpty,
  isEmptyObject,
  isFocusable,
  isFunction,
  isGray,
  isHTMLElement,
  isHidden,
  isLight,
  isMac,
  isMouseEvent,
  isMultiTouchEvent,
  isNotNumber,
  isNull,
  isNumber,
  isNumeric,
  isObject,
  isRefObject,
  isSafari,
  isSomeElement,
  isString,
  isTabbable,
  isTone,
  isTouchDevice,
  isTouchEvent,
  isUndefined,
  isUnit,
  isValidElement,
  keysFormObject,
  lightenColor,
  memoizeObject,
  merge,
  mergeRefs,
  noop,
  objectFromEntries,
  omitChildren,
  omitObject,
  parseToHsla,
  parseToHsv,
  parseToRgba,
  percentToValue,
  pickChildren,
  pickObject,
  platform,
  pointFromMouse,
  pointFromTouch,
  randomColor,
  replaceObject,
  rgbaTo,
  roundNumberToStep,
  runIfFunc,
  sameColor,
  shadeColor,
  splitObject,
  tintColor,
  toCamelCase,
  toKebabCase,
  toNumber,
  toPascalCase,
  toPrecision,
  toTitleCase,
  transformSize,
  transparentizeColor,
  useAsync,
  useAsyncFunc,
  useAsyncRetry,
  useCallbackRef,
  useIsMounted,
  useMergeRefs,
  useSafeLayoutEffect,
  useUnmountEffect,
  useUpdateEffect,
  valueToPercent,
  vendor
};
