// @yamada-ui/ripple/dist/index.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Ripple: () => Ripple,
  useRipple: () => useRipple
});
module.exports = __toCommonJS(src_exports);

// src/ripple.tsx
var import_motion = require("@yamada-ui/motion");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var clamp = (value, min, max) => Math.min(Math.max(value, min), max);
var Ripple = ({
  className,
  style,
  color = "currentColor",
  isDisabled,
  ripples,
  onAnimationComplete,
  onClear,
  ...rest
}) => {
  if (isDisabled) return null;
  const css = {
    rounded: "fallback(full, 9999px)",
    zIndex: "fallback(kurillin, 9)"
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: ripples.map(({ key, size, x, y }) => {
    const duration = clamp(0.01 * size, 0.2, size > 100 ? 0.75 : 0.5);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_motion.AnimatePresence, { mode: "popLayout", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_motion.motion.span,
      {
        className: (0, import_utils.cx)("ui-ripple", className),
        style: {
          height: `${size}px`,
          left: x,
          pointerEvents: "none",
          position: "absolute",
          top: y,
          transformOrigin: "center",
          width: `${size}px`,
          ...style
        },
        animate: { opacity: 0, transform: "scale(2)" },
        bgColor: color,
        exit: { opacity: 0 },
        initial: { opacity: 0.35, transform: "scale(0)" },
        transition: { duration },
        __css: css,
        ...rest,
        onAnimationComplete: (0, import_utils.handlerAll)(
          onAnimationComplete,
          () => onClear(key)
        )
      }
    ) }, key);
  }) });
};
Ripple.displayName = "Ripple";
Ripple.__ui__ = "Ripple";

// src/use-ripple.ts
var import_utils2 = require("@yamada-ui/utils");
var import_react = require("react");
var useRipple = ({
  disabled,
  isDisabled,
  ...rest
} = {}) => {
  const [ripples, setRipples] = (0, import_react.useState)([]);
  const onPointerDown = (0, import_react.useCallback)(
    (ev) => {
      if (disabled || isDisabled) return setRipples([]);
      const trigger = ev.currentTarget;
      const size = Math.max(trigger.clientWidth, trigger.clientHeight);
      const rect = trigger.getBoundingClientRect();
      setRipples((prev) => [
        ...prev,
        {
          key: (0, import_utils2.createId)(prev.length.toString()),
          size,
          x: ev.clientX - rect.x - size / 2,
          y: ev.clientY - rect.y - size / 2
        }
      ]);
    },
    [disabled, isDisabled]
  );
  const onClear = (0, import_react.useCallback)((key) => {
    setRipples((prev) => prev.filter((item) => item.key !== key));
  }, []);
  return {
    ripples,
    onClear,
    onPointerDown: (0, import_utils2.handlerAll)(onPointerDown, rest.onPointerDown)
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Ripple,
  useRipple
});
