// @yamada-ui/motion/dist/index.mjs
"use client"
import {
  motion
} from "./chunk-4MMUMOXF.mjs";
import {
  Motion
} from "./chunk-TEUF5ZAQ.mjs";
import {
  motionForwardRef
} from "./chunk-RXCU3ZFK.mjs";
import {
  MOTION_TRANSITION_DEFAULTS,
  MOTION_TRANSITION_EASINGS,
  MOTION_TRANSITION_VARIANTS,
  transitionEnter,
  transitionExit
} from "./chunk-AWDPDZLD.mjs";

// src/index.ts
import {
  AcceleratedAnimation,
  addPointerInfo,
  addScaleCorrector,
  animate,
  AnimatePresence,
  AnimateSharedLayout,
  animateValue,
  animateVisualElement,
  animationControls,
  animations,
  anticipate,
  backIn,
  backInOut,
  backOut,
  buildTransform,
  calcLength,
  cancelFrame,
  cancelSync,
  circIn,
  circInOut,
  circOut,
  clamp,
  color,
  complex,
  createBox,
  createDomMotionComponent,
  createMotionComponent,
  createScopedAnimate,
  cubicBezier,
  delay,
  DeprecatedLayoutGroupContext,
  disableInstantTransitions,
  distance,
  distance2D,
  domAnimation,
  domMax,
  DragControls,
  easeIn,
  easeInOut,
  easeOut,
  filterProps,
  FlatTree,
  frame,
  frameData,
  interpolate,
  invariant,
  inView,
  isBrowser,
  isDragActive,
  isMotionComponent,
  isMotionValue,
  isValidMotionProp,
  LayoutGroup,
  LayoutGroupContext,
  LazyMotion,
  m,
  makeUseVisualState,
  mirrorEasing,
  mix,
  MotionConfig,
  MotionConfigContext,
  MotionContext,
  MotionGlobalConfig,
  motionValue,
  optimizedAppearDataAttribute,
  pipe,
  PresenceContext,
  progress,
  px,
  Reorder,
  resolveMotionValue,
  reverseEasing,
  scroll,
  scrollInfo,
  spring,
  stagger,
  startOptimizedAppearAnimation,
  steps,
  SwitchLayoutGroupContext,
  sync,
  transform,
  unwrapMotionComponent,
  useAnimate,
  useAnimation,
  useAnimationControls,
  useAnimationFrame,
  useCycle,
  useDeprecatedAnimatedState,
  useDeprecatedInvertedScale,
  useDomEvent,
  useDragControls,
  useElementScroll,
  useForceUpdate,
  useInstantLayoutTransition,
  useInstantTransition,
  useInView,
  useIsomorphicLayoutEffect,
  useIsPresent,
  useMotionTemplate,
  useMotionValue,
  useMotionValueEvent,
  usePresence,
  useReducedMotion,
  useReducedMotionConfig,
  useResetProjection,
  useScroll,
  useSpring,
  useTime,
  useTransform,
  useVelocity,
  useViewportScroll,
  useWillChange,
  VisualElement,
  visualElementStore,
  warning,
  wrap
} from "framer-motion";
export {
  AcceleratedAnimation,
  AnimatePresence,
  AnimateSharedLayout,
  DeprecatedLayoutGroupContext,
  DragControls,
  FlatTree,
  LayoutGroup,
  LayoutGroupContext,
  LazyMotion,
  MOTION_TRANSITION_DEFAULTS,
  MOTION_TRANSITION_EASINGS,
  MOTION_TRANSITION_VARIANTS,
  Motion,
  MotionConfig,
  MotionConfigContext,
  MotionContext,
  MotionGlobalConfig,
  Reorder as MotionReorder,
  PresenceContext,
  SwitchLayoutGroupContext,
  VisualElement,
  addPointerInfo,
  addScaleCorrector,
  animate,
  animateValue,
  animateVisualElement,
  animationControls,
  animations,
  anticipate,
  backIn,
  backInOut,
  backOut,
  buildTransform,
  calcLength,
  cancelFrame,
  cancelSync,
  circIn,
  circInOut,
  circOut,
  clamp,
  color,
  complex,
  createBox,
  createDomMotionComponent,
  createMotionComponent,
  createScopedAnimate,
  cubicBezier,
  delay,
  disableInstantTransitions,
  distance,
  distance2D,
  domAnimation,
  domMax,
  easeIn,
  easeInOut,
  easeOut,
  filterProps,
  frame,
  frameData,
  inView,
  interpolate,
  invariant,
  isBrowser,
  isDragActive,
  isMotionComponent,
  isMotionValue,
  isValidMotionProp,
  m,
  makeUseVisualState,
  mirrorEasing,
  mix,
  motion,
  motionForwardRef,
  motionValue,
  optimizedAppearDataAttribute,
  pipe,
  progress,
  px,
  resolveMotionValue,
  reverseEasing,
  scroll,
  scrollInfo,
  spring,
  stagger,
  startOptimizedAppearAnimation,
  steps,
  sync,
  transform,
  transitionEnter,
  transitionExit,
  unwrapMotionComponent,
  useAnimate,
  useAnimationControls,
  useAnimationFrame,
  useCycle,
  useDeprecatedAnimatedState,
  useDeprecatedInvertedScale,
  useDomEvent,
  useDragControls,
  useElementScroll,
  useForceUpdate,
  useInView,
  useInstantLayoutTransition,
  useInstantTransition,
  useIsPresent,
  useIsomorphicLayoutEffect,
  useAnimation as useMotionAnimation,
  useMotionTemplate,
  useMotionValue,
  useMotionValueEvent,
  usePresence,
  useReducedMotion,
  useReducedMotionConfig,
  useResetProjection,
  useScroll,
  useSpring,
  useTime,
  useTransform,
  useVelocity,
  useViewportScroll,
  useWillChange,
  visualElementStore,
  warning,
  wrap
};
